added documentation for public parts of the crate

serejkus opened the pull request:
Added docs for public parts of the crate, as requested in #17 and #23.

English is not my native language, so feel free to fix grammar issues. The same applies to examples: most of them are just copy-paste and involve parsing of semver structs from strings, maybe there might be a better approach in some cases.

Commit:
added documentation for public parts of the crate

serejkus commented:
@KodrAus, any suggestions on this PR?

KodrAus commented:
Sorry I somehow missed this. Thanks for the ping @serejkus and for taking on these docs! I'll dig in to this later today.

KodrAus commented on src/lib.rs:
```diff
@@ -1,3 +1,52 @@
+//! collection of [`range::Predicate`]
+//! representing data for version comparision.
+//!
+//! # Example
```
Let's change this to _Examples_ to be consistent.

steveklabnik replied:
In general, all of them should always be `Examples` :)

KodrAus commented on src/lib.rs:
```diff
@@ -1,3 +1,52 @@
+//! Collection of structures and helper functions for parsing semantic version.
+//!
+//! This crate contains data structures for holding version data and comparision of versions
```
Typo here: _comparision_ to _comparison_. I always type comparision for some reason...

Looks like there are a few other places we should replace _comparision_ with _comparison_.

KodrAus commented on src/lib.rs:
```diff
@@ -1,3 +1,52 @@
+//! use semver_parser::version;
+//!
+//! # fn try_main() -> Result<(), String> {
+//!
```
We should remove these newlines after hidden lines, it makes the code here look a little funny, but when it renders in the docs you only have 1 visible newline.

Does that make sense?

KodrAus commented on src/lib.rs:
````diff
@@ -1,3 +1,52 @@
+//!
+//! # Example
+//!
+//! ```rust
````
For the docs we don't actually need the `rust` annotation here. `rustdoc` assumes the code block contains testable Rust by default.

KodrAus commented on src/range.rs:
```diff
@@ -1,30 +1,222 @@
+//! Version range and requirements data and functions (used for version comparision)
```
Let's throw a `.` at the end of these introduction lines in all the modules, structures and methods to be consistent.

KodrAus commented on src/version.rs:
````diff
@@ -1,19 +1,108 @@
+/// ```
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Version {
+    /// major version as number (`0` in `"0.1.2"`);
````
Let's capitalise the _major_, _minor_ etc on these docs and use a `.` instead of a `;` at the end.

KodrAus reviewed (commented):
Thanks @serejkus this is an awesome effort! I've just left some comments for a coupe of typos.

serejkus commented:
Thanks for pointing out! Will try to fix those issues in a day.

steveklabnik commented:
Seconded, thanks a ton @serejkus :)

Commit:
fixing docs: consistent style, spelling fixes, links to cargo docs added

serejkus commented:
Fixed issues with style and added links to cargo docs explaining caret and tilde version requirements. Two commits are ok or is it better to rebase it into a single commit?

steveklabnik commented:
Two commits is just fine. Thanks again!
